### PR TITLE
Fix factors crash in `tm_g_association`

### LINKED ISSUE #645: factors crash `tm_g_association`
In `tm_g_association`, when a selected categorical variable is a factor (not character), the app crashes.

##### error message
```
Listening on http://127.0.0.1:6444
[INFO] 2024-02-19 15:28:58.7853 pid:71573 token:[29b46c16] teal Initializing reporter_previewer_module
Warning: Error in match.arg: 'arg' must be of length 1
  143: stop
  142: match.arg
  141: bivariate_ggplot_call [/home/users/chlebowa/NEST/teal.modules.general/R/tm_g_bivariate.R#750]
  140: bivariate_plot_call [/home/users/chlebowa/NEST/teal.modules.general/R/tm_g_bivariate.R#688]
  139: <reactive> [/home/users/chlebowa/NEST/teal.modules.general/R/tm_g_association.R#308]
  123: output_q
  122: <reactive> [/home/users/chlebowa/NEST/teal.modules.general/R/tm_g_association.R#425]
  106: plot_r
   97: renderUI [/home/users/chlebowa/NEST/teal.widgets/R/plot_with_settings.R#412]
   96: func
   83: renderFunc
   82: output$teal-main_ui-root-association-module-myplot-plot_out_main
    1: runApp
Warning: Error in match.arg: 'arg' must be of length 1
  102: <Anonymous>
  101: stop
  100: output_q
   97: renderText [/home/users/chlebowa/NEST/teal.modules.general/R/tm_g_association.R#436]
   96: func
   83: renderFunc
   82: output$teal-main_ui-root-association-module-title
    1: runApp
Warning: Error in match.arg: 'arg' must be of length 1
  1: runApp
```

##### this app fails:
```
# general data exapmle
library(teal.modules.general)
library(teal.widgets)

data <- teal_data()
data <- within(data, {
  library(nestcolor)
  CO2 <- as.data.frame(CO2)
})
datanames(data) <- c("CO2")

app <- init(
  data = data,
  modules = modules(
    tm_g_association(
      ref = data_extract_spec(
        dataname = "CO2",
        select = select_spec(
          label = "Select variable:",
          choices = variable_choices(data[["CO2"]], c("Plant", "Type", "Treatment")),
          selected = "Plant",
          fixed = FALSE
        )
      ),
      vars = data_extract_spec(
        dataname = "CO2",
        select = select_spec(
          label = "Select variables:",
          choices = variable_choices(data[["CO2"]], c("Plant", "Type", "Treatment")),
          selected = "Treatment",
          multiple = TRUE,
          fixed = FALSE
        )
      ),
      ggplot2_args = ggplot2_args(
        labs = list(subtitle = "Plot generated by Association Module")
      )
    )
  )
)
runApp(app, launch.browser = TRUE)

```

##### this app works
```
# general data exapmle
library(teal.modules.general)
library(teal.widgets)

data <- teal_data()
data <- within(data, {
  library(nestcolor)
  CO2 <- as.data.frame(CO2)
  factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
  CO2[factors] <- lapply(CO2[factors], as.character)
})
datanames(data) <- c("CO2")

app <- init(
  data = data,
  modules = modules(
    tm_g_association(
      ref = data_extract_spec(
        dataname = "CO2",
        select = select_spec(
          label = "Select variable:",
          choices = variable_choices(data[["CO2"]], c("Plant", "Type", "Treatment")),
          selected = "Plant",
          fixed = FALSE
        )
      ),
      vars = data_extract_spec(
        dataname = "CO2",
        select = select_spec(
          label = "Select variables:",
          choices = variable_choices(data[["CO2"]], c("Plant", "Type", "Treatment")),
          selected = "Treatment",
          multiple = TRUE,
          fixed = FALSE
        )
      ),
      ggplot2_args = ggplot2_args(
        labs = list(subtitle = "Plot generated by Association Module")
      )
    )
  )
)
runApp(app, launch.browser = TRUE)
```

### PR DESCRIPTION
**Pull request by vedhav**
Closes #645

Changes:
Pass the base class into `bivariate_plot_call` because we only expect the base class in the downstream function call.

### CODE REVIEW

**Review by chlebowa — COMMENTED** [review]
The fix does work but I am not satisfied with the reasoning.
Usually one would use the most junior class by taking `ref_class <- class(ANL[[ref_name]])[1L]`. The core of the problem is that in this case the junior class, `ordered`, is not supported. `factor`, however, is. 

Perhaps a better way would be to drop the `ordered` class. What do you think?

**Comment by vedhav** [comment]
> > I disagree that `ordered` is not supported
> 
> image 😉
> 
> > we do a `match.arg`
> 
> where?

😲 When I tested it worked all the time. Can you tell me how to reproduce this?

I'll convert `ordered` to just `factor`.

**Comment by chlebowa** [comment]
I just ran the example app on your branch but with `ref_class <- class(ANL[[ref_name]])[1L]`.

<details>
<summary>app code</summary>
```
# general data exapmle
devtools::load_all()
library(teal.widgets)

data <- teal_data()
data <- within(data, {
  library(nestcolor)
  CO2 <- as.data.frame(CO2)
  CO2$Treatment <- seq(Sys.time(), by = 1, length.out = nrow(CO2))
})
datanames(data) <- c("CO2")

app <- init(
  data = data,
  modules = modules(
    tm_g_association(
      ref = data_extract_spec(
        dataname = "CO2",
        select = select_spec(
          label = "Select variable:",
          choices = variable_choices(data[["CO2"]], c("Plant", "Type", "Treatment")),
          selected = "Plant",
          fixed = FALSE
        )
      ),
      vars = data_extract_spec(
        dataname = "CO2",
        select = select_spec(
          label = "Select variables:",
          choices = variable_choices(data[["CO2"]], c("Plant", "Type", "Treatment")),
          selected = "Treatment",
          multiple = TRUE,
          fixed = FALSE
        )
      ),
      ggplot2_args = ggplot2_args(
        labs = list(subtitle = "Plot generated by Association Module")
      )
    )
  )
)
runApp(app, launch.browser = TRUE)
```
</details>

File: R/tm_g_association.R [hunk 1]
```diff
@@ -336,7 +336,7 @@ srv_tm_g_association <- function(id,
       teal::validate_has_data(ANL[, c(ref_name, vars_names)], 3, complete = TRUE, allow_inf = FALSE)
 
       # reference
-      ref_class <- class(ANL[[ref_name]])
+      ref_class <- class(ANL[[ref_name]])[1]
```
Comment (chlebowa) [inline R/tm_g_association.R:339]:
```suggestion
      ref_class <- class(ANL[[ref_name]])[1L]
```

File: R/tm_g_association.R [hunk 2]
```diff
@@ -373,7 +373,7 @@ srv_tm_g_association <- function(id,
       print_call <- quote(print(p))
 
       var_calls <- lapply(vars_names, function(var_i) {
-        var_class <- class(ANL[[var_i]])
+        var_class <- class(ANL[[var_i]])[1]
```
Comment (chlebowa) [inline R/tm_g_association.R:376]:
```suggestion
        var_class <- class(ANL[[var_i]])[1L]
```

File: R/tm_g_bivariate.R [hunk 3]
```diff
@@ -797,8 +797,8 @@ substitute_q <- function(x, env) {
 #' bivariate_ggplot_call("numeric", "factor")
 #' bivariate_ggplot_call("factor", "numeric")
 #' bivariate_ggplot_call("factor", "factor")
-bivariate_ggplot_call <- function(x_class = c("NULL", "numeric", "integer", "factor", "character", "logical"),
-                                  y_class = c("NULL", "numeric", "integer", "factor", "character", "logical"),
+bivariate_ggplot_call <- function(x_class = c("NULL", "numeric", "integer", "factor", "ordered", "character", "logical"), # nolint: line_length
```
Comment (chlebowa) [inline R/tm_g_bivariate.R:800]:
```suggestion
bivariate_ggplot_call <- function(x_class = c("NULL", "numeric", "integer", "factor", "ordered", "character", "logical"), # nolint: line_length.
```

File: R/tm_g_bivariate.R [hunk 4]
```diff
@@ -797,8 +797,8 @@ substitute_q <- function(x, env) {
...
+                                  y_class = c("NULL", "numeric", "integer", "factor", "ordered", "character", "logical"), # nolint: line_length
```
Comment (chlebowa) [inline R/tm_g_bivariate.R:801]:
```suggestion
                                  y_class = c("NULL", "numeric", "integer", "factor", "ordered", "character", "logical"), # nolint: line_length.
```

**Comment by vedhav** [comment]
Thanks for the example. I see that this is just because when you made the change to pass the `ordered` you did not make the required validation changes to support it, Once we do this it should start working fine. I made this change in the commit 
[77519dd](https://github.com/insightsengineering/teal.modules.general/pull/692/commits/77519ddb1a1c40a2fcbd0fa975727f202a03024b)
So I still believe that: ordered is supported in the downstream code, technically it is not supported because we validated it out, but it should be.

File: R/tm_g_bivariate.R (hunk 4 above)
Comment (chlebowa) [reply R/tm_g_bivariate.R:801]: How about removing the default value of `x_class` and having this inside?
```
x_class <- switch(
  x_class,
  "integer" =, "numeric" = "numeric",
  "logical" =, "character" =, "factor" =, "ordered" = "factor",
  stop("unsupported x_class: ", toString(x_class))
)
```
If you add `if (is.null(x_class) && is.null(y_class)) stop("either x or y is required")`, everything will be covered.

**Review by chlebowa — CHANGES_REQUESTED** [review]

**Comment by chlebowa** [comment]
Forgot to add review comment: requesting changes because of missing full stops in linter exceptions.
Do consider the other suggestions, though :slightly_smiling_face:

File: R/tm_g_bivariate.R (hunk 4 above)
Comment (vedhav) [reply R/tm_g_bivariate.R:801]: I like this approach 👍🏽 Much better than `match.args`

**Comment by chlebowa** [comment]
You sure you don't want the more concise syntax? `"logical" =, "character" =, "factor" =, "ordered" = "factor"`

File: R/tm_g_association.R (hunk 1 above)
Comment (vedhav) [reply R/tm_g_association.R:339]: What difference does it make?  Indexing a small vector has no difference when it comes to what we use. I'd prefer to use just the `1` as it's easier on the eyes.
We use `[1]` in over 100 places in the repo. And a couple of places where we use `[1L]`.

File: R/tm_g_association.R (hunk 2 above)
Comment (chlebowa) [reply R/tm_g_association.R:376]: pet peeve, feel free to ignore

**Review by chlebowa — APPROVED** [review]
:+1:

**Comment by vedhav** [comment]
Just to confirm, Merging this into `main` is okay right? Or should I rebase it to pre-release branch and merge it there?

**Pull request merged**